SOAP catalogue

bwvdnbro opened the pull request:
This PR adds support for SOAP catalogues. SOAP (https://github.com/SWIFTSIM/SOAP) is a new tool to calculate (sub)halo properties. It uses the subhalo membership information from VELOCIraptor and its centre of potential to calculate a large set of additional properties, like subhalo masses, star formation rates, spherical overdensity radii... These are output as a VR-like catalogue file with a much clearer structure and unit metadata, similar to the SWIFT snapshot metadata.

While the new SOAP output structure would make a new SOAP reader module desirable, this would make it hard to convert from VR to SOAP in the pipeline, since that depends a lot on functionality in this module. As a (temporary?) solution, this PR instead makes it possible to replace the VR catalogue with a SOAP catalogue with minimal changes to the API used in the pipeline. The pipeline can then slowly be adapted to use the new SOAP variable names, without the need for an immediate major update.

In practice, this PR:
 - defines a more general `Catalogue` interface of which the existing `VelociraptorCatalogue` is one child class. A new `SOAPCatalogue` child class is defined to handle SOAP catalogues. `load()` uses the appropriate child class using a `try..except` block.
 - defines a general interface for other parts of the module and external code to interact with the catalogue, the most important part of which is `catalogue.get_quantity()`, which should be used to get a dataset where you would use a catalogue attribute before. This is the only change to the external API, and will/does affect code in the pipeline registration functions.
 - defines a new translator function that can convert old VR dataset names to new SOAP dataset names, taking into account the fact that SOAP, unlike VR, supports multidimensional datasets. Because of this translator function, it is possible to use both the VR name and the SOAP name for the same dataset.

Apart from these changes, the PR also contains some new debugging and testing code.

The PR is still missing functionality to obtain subhalo membership information from SOAP catalogues, as is possible for VR catalogues. This can be added later but is not a priority right now.

@JBorrow I imagine that you might have some thoughts on the new design. I originally tried to stick to the original attribute-based access to the catalogue datasets, but in the end having the `catalogue.get_quantity()` function was pretty much inevitable to deal with having multiple names for the same dataset. The SOAP output file structure maps a lot better to multi-level attributes (e.g. "SO/200_crit/TotalMass" --> `so.200_crit.totalmass`), but having a mapping that works for both SOAP and VR at the same time is pretty much impossible. We could go back to attribute-based access once the entire pipeline uses the new SOAP names, but that will take a while.

@EvgeniiChaikin I hope we can get this merged soon, but in case that does not happen, you can already use this branch to test the SOAP-based pipeline. I will create an appropriate PR in the pipeline-configs repository later this week.

bwvdnbro commented:
Thanks! It would be good if this could be processed before I leave Leiden (in exactly 3 weeks). I would like to keep @EvgeniiChaikin in the loop, just in case there is more work to do past that point.

EvgeniiChaikin commented:
@bwvdnbro I have already started looking into this PR and testing it. It will probably take me a couple of more days before I come up with the first comments.

But I want to say that the PR already looks very nice and well-documented so the review should not take long.

EvgeniiChaikin reviewed (changes requested):
Hi @bwvdnbro,

I have left small comments, mostly in `soap_catalogue.py`.

**More general comments:**

- Many functions in `soap_catalogue.py` do not have type hints. I'd add more type hints, if possible.
- We should also change `VelociraptorCatalogue` -> `Catalogue` in `velociraptor/fitting_formulae/smhmr.py`. 
- I am fine with the new `get_quantity` method. Regarding @JBorrow's comment, if really needed, I think it should be possible to preserve the attribute-style access (including the currently used names of the fields) with the help of `translator.py`. 
- By the way, if we decide to stick with the `get_quantity` method, I'd suggest doing a bit of restructuring. Namely, since we now have two equally important child classes, `SOAPCatalogue` and `VelociraptorCatalogue`, I think they should have their own independent implementations of the `get_quantity` method. At the moment,  the  `get_quantity` method of the `VelociraptorCatalogue` class is defined in the parent class `Catalogue`, while `get_quantity` of the `SOAPCatalogue` class is defined within the `SOAPCatalogue` class itself.
- I have also found that some plots have different axis labels if the plotted fields were produced by SOAP. E.g. in
![stellar_mass_star_fe_over_h_mol_lofloor_50](https://user-images.githubusercontent.com/20153933/206425462-34f9f6bf-5366-4c31-b6bb-07f3c39c92df.png)

I don't think we want to have `ExclusiveSphere` in the x axis's label. I know that you have thought of it and this is a mechanism to handle the labels (I guess by using `get_particle_property_name_conversion()` in `translator.py`?). What is not clear to me is why in some cases the names that we specify in `registration.py` do not propagate to figures' labels.


PS _This PR is rather large, so I will soon have another look at it and try to come up with a new round of comments._

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+
+class VelocityDispersion(CatalogueDerivedDataset):
+    def compute_value(self, velocity_dispersion_matrix):
+        return np.sqrt(
+            velocity_dispersion_matrix[:, 0]
+            + velocity_dispersion_matrix[:, 1]
+            + velocity_dispersion_matrix[:, 2]
+        )
```
This is incorrect. `velocity_dispersion_matrix` is a 6 element array and we only want to sum the first 3 elements. But you are right that this can be done in a single line.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+                self.cosmology = FlatLambdaCDM(H0=H0, Om0=Omega_m, Ob0=Omega_b)
+            try:
+                boxsize = handle["SWIFT/Header"].attrs["BoxSize"][0]
+            except:
```
The `SWIFT/Header` group did not exist when I first started implementing SOAP support, so this was just to deal with older catalogue files. This whole block can be removed.

bwvdnbro commented:
> I don't think we want to have `ExclusiveSphere` in the x axis's label. I know that you have thought of it and this is a mechanism to handle the labels (I guess by using `get_particle_property_name_conversion()` in `translator.py`?). What is not clear to me is why in some cases the names that we specify in `registration.py` do not propagate to figures' labels.

Which name shows up in the label depends on the order in which the various curves are added. `unyt.matplotlib_support` adds the first label it gets. If you first plot observational data, it will take the label from those. If you first plot the simulation, it will use whatever label was added to the `name`.

I have thought about this and I would really like to avoid using a translator function that provides a custom label for every dataset, because that is very hard to maintain. But it has also turned out to be pretty hard to come up with a generic way to convert the dataset name into a label (which is what is happening now). I actually like having the full dataset path, since then it is immediately clear what SOAP quantity we use: in this case the exclusive 3D aperture with a 50 kpc radius. If you omit the "ExclusiveSphere", then it is not clear whether you use a projected or a 3D aperture, and whether or not unbound particles are excluded or not. Omitting the 50 kpc is even more confusing. You would not put this in a paper, but for a pipeline plot that seems pretty useful to me.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+            w0 = cosmology["w_0"][0]
+            Omega_b = cosmology["Omega_b"][0]
+            if w0 != -1.0:
+                self.cosmology = wCDM(
+                    H0=H0, Om0=Omega_m, Ode0=Omega_DE, w0=w_of_DE, Ob0=Omega_b
+                )
```
Good point. I clearly haven't tested that case.

bwvdnbro replied:
Has that discussion now been resolved completely?

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+
+class SOAPCatalogue(Catalogue):
+
+    file_name: str
+    root: CatalogueGroup
+    names_used: Set[str]
```
I clearly don't understand Python myself. I thought this was just a way of declaring class variables before they are actually initialised (like you would do in C++). I did not know these declarations actually have a different meaning. If that is the case, then these declarations should go.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+        self._value = value
+
+    def del_value(self, group):
+        del self._value
```
The `group` thing makes sense to me now. When we access the `property`, we are actually doing
```
getattr(<CatalogueGroup instance>, "name")
```
which is then translated into
```
<property getter for "name">(<CatalogueGroup instance>)
  --> <CatalogueDataset instance>.get_value(<CatalogueGroup instance>)
  --> get_value(<CatalogueDataset instance "self">, <CatalogueGroup instance>)
```
That means we need the `get_value(self, group)` syntax, even though `group` is never used.

bwvdnbro replied:
I think I disagree about moving `_value` and `set_value`, `del_value` to `CatalogueElement`. `CatalogueGroup` also inherits from this class and those objects do not need this variable and these functions. So that would not be very clean.

To avoid code duplication, I could introduce another class that inherits from `CatalogueElement` and that serves as parent class for `CatalogueDataset` and `CatalogueDerivedDataset`, but that would likely introduce more code than it saves.

JBorrow commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+        self.names_used.add(quantity_name)
+        return value
+
+    def get_quantity(self, quantity_name):
```
Can we get some documentation for this method and why it is required?

JBorrow commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+                return quantity[:, colidx]
+            else:
+                return quantity
+        except NotImplementedError as err:
+            if quantity_name in [
+                "apertures.veldisp_star_10_kpc",
+                "apertures.veldisp_star_30_kpc",
+            ]:
+                print(f"Ignoring missing {quantity_name}...")
+                return None
+            raise err
```
What is going on here? Why this special case? Please document and explain. Print statements for these things are generally a bad idea, too. You should raise a Warning, maybe?

JBorrow commented on velociraptor/catalogue/translator.py:
```diff
@@ -8,6 +8,1396 @@
 from velociraptor.units import VelociraptorUnits
 
 
+def VR_to_SOAP(particle_property_name: str):
```
What does this return? Please add a type hint.

JBorrow commented on velociraptor/catalogue/translator.py:
```diff
@@ -8,6 +8,1396 @@
 
 
+def VR_to_SOAP(particle_property_name: str):
+    """
```
Oh my goodness, I hate it. Though I understand why it is needed for now. We need to make sure this remains temporary.

bwvdnbro replied:
I totally agree.

JBorrow commented on velociraptor/observations/objects.py:
```diff
@@ -647,14 +647,18 @@ def plot_on_axes(self, axes: Axes, errorbar_kwargs: Union[dict, None] = None):
+                **kwargs,
+                label=data_label,
+            )
+        except ValueError as e:
+            print(data_label, self.x, self.y, self.x_scatter, self.y_scatter)
```
Please document what is going on here and why.

JBorrow reviewed (commented):
Many comments, this needs a lot more documentation. The overall structure looks okay though.

bwvdnbro commented on velociraptor/__init__.py:
```diff
@@ -73,14 +74,17 @@ def load(
-    catalogue = VelociraptorCatalogue(
-        filename, disregard_units=disregard_units, mask=mask
-    )
+    try:
+        catalogue = VelociraptorCatalogue(
+            filename, disregard_units=disregard_units, mask=mask
+        )
+    except KeyError:
+        catalogue = SOAPCatalogue(filename)
```
I disagree. There are many ways to check whether a file is a SOAP or VR output, but all of those would require specific knowledge about the structure of the output file. Having a `SOAP` header attribute would not be any different. So sure, we could hard-code such a check and then update it if the structure of the output file were to change at some point in the future. But is that really any different from trying to open the file as a VR catalogue and using SOAP as a fallback when that fails? I don't see any particular overhead in doing that, and then at least the knowledge of the catalogue structure stays limited to that particular class.

bwvdnbro replied:
That is a good point, I will use a custom exception instead.

bwvdnbro commented on velociraptor/autoplotter/objects.py:
```diff
@@ -801,12 +800,14 @@ def get_quantity_from_catalogue_with_mask(
                 )
             self.structure_mask = logical_and(
                 self.structure_mask,
-                catalogue.structure_type.structuretype == self.select_structure_type,
+                catalogue.get_quantity("structure_type.structuretype")
+                == self.select_structure_type,
```
Yes, this works. In the SOAP catalogue, this dataset is called `vr.structuretype`.

bwvdnbro commented on velociraptor/autoplotter/objects.py:
```diff
@@ -1217,6 +1208,17 @@ def create_plots(
                     import sys, traceback
 
+                    _, _, exc_traceback = sys.exc_info()
+                    print("Traceback:")
```
Sure, but this is just copying the bit of code just below, so this is not really part of this PR.

bwvdnbro replied:
Also: the traceback functions that are used in this block and the other blocks explicitly print to `stdout`. So that does not seem very consistent with your comment.

JBorrow replied:
Yeah they do but I think that may have been a mistake. I think we want the errors to go to stderr, and then the users can redirect as they wish.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+                )
+
+    def _register_extra_properties(self):
+        try:
+            stellar_velocity_dispersion_matrix = self.properties[
+                "stellarvelocitydispersionmatrix"
+            ][0]
+            self.elements.append(
+                VelocityDispersion(
+                    self.file_name,
+                    "stellarvelocitydispersion",
+                    [stellar_velocity_dispersion_matrix],
+                )
+            )
+            el = self.elements[-1]
```
Why this is not done in SOAP: because the 1D velocity dispersion can trivially be derived from the velocity dispersion matrix output by SOAP, and we did not want to store any more data in the SOAP output than strictly necessary.

Why this is done here: because VR did output a 1D velocity dispersion, so this is the only way to make sure we can replace VR with SOAP without changing the pipeline. We could also update the pipeline and turn this into a registration function.

Although I must say that I prefer this way over a registration function, since this way will still use lazy reading.

JBorrow replied:
I just worry about the potential debt that this function in particular creates... It is a very odd piece of code to have here. I think a registration function may be best? Are you still using registration functions for other things?

bwvdnbro replied:
We do use registration functions for other things, but the problem is that quantities created by registration functions have a different name in the autoplotter. So that would require a change to the pipeline. I was trying to avoid that at this point.

I also personally think this way of doing things has some potential, since it allows you to define derived quantities without necessarily reading the data and computing them. There are tons of registered quantities in the COLIBRE pipeline that are read and computed but never actually used. But I agree that this might not be the cleanest mechanism to create the derived datasets. I have no time left to come up with a better mechanism, but I know @EvgeniiChaikin was interested in exploring this further.

So my proposal is to keep this for now, but aim to remove/change this mechanism in the future, either by using a registration function and updating the pipeline, or by expanding the use of derived datasets through a cleaner mechanism.

JBorrow replied:
Ok, we can do that.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+        with h5py.File(self.file_name, "r") as handle:
+            self.root = CatalogueGroup(self.file_name, "", handle)
+            cosmology = handle["SWIFT/Cosmology"].attrs
+            self.units = SimpleNamespace()
```
Yes, `catalogue.units` is required for compatibility with the old VR API. The pipeline uses things like `catalogue.units.box_length`. I am sure this can be changed to something more elegant in the future, but for now it serves its purpose.

JBorrow replied:
I mean for choosing SimpleNamespace instead of using a dataclass or a named tuple. Structured records like our units should use one of these instead of a SimpleNamespace.

bwvdnbro replied:
I don't know what dataclasses or named tuples are. Could you tell me how you would implement this?

JBorrow replied:
Well for the VR side we have a custom object that is the units: https://velociraptor-python.readthedocs.io/en/latest/_modules/velociraptor/units.html#VelociraptorUnits.

You could have SOAPUnits which is a class containing some of this information.

A dataclass is just a class but has less gubbins. It's basically a struct.

A namedtuple is even lighter weight, but it is immutable once created.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+                boxsize = handle["SWIFT/Header"].attrs["BoxSize"][0]
+            except:
+                boxsize = 1000.0
+            boxsize *= unyt.Mpc
```
Yes, of course. They were not at the time I hard-coded the box size, but now they are and I am reading them from the file.

bwvdnbro commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -0,0 +1,266 @@
+        self._register_extra_properties()
+
+    def _register_elements(self, handle):
+        h5group = handle[self.name] if self.name != "" else handle["/"]
```
This actually does not work, since we are later on appending to the name. So `""` is the only value that makes sense.

bwvdnbro commented:
I think I have addressed most of the comments. As soon as the cosmology fix for swiftsimio is confirmed, I will port that over. The other discussions can keep going, but I can't promise I will have any time left to address them.

MatthieuSchaller commented:
Regarding the cosmology thing above, the latest `swiftsimio` can now correctly handle hings.

JBorrow commented:
@EvgeniiChaikin do you have time to include these changes in SOAP?

EvgeniiChaikin commented:
@JBorrow yes, I will be able to have a look in the coming days.

JBorrow commented on velociraptor/catalogue/soap_catalogue.py:
```diff
@@ -12,13 +11,63 @@
 from pathlib import Path
 
 
+class SWIFTUnits_from_dictionary(object):
```
I think this name is really weird, on a second look. The mixing of camel and snake case is really odd, and it's hard to tell if this is a function or a class.

How about `SWIFTUnitsMockup`?

JBorrow commented:
Are we ready to go on this now?

MatthieuSchaller commented:
Can we maybe test it first? @EvgeniiChaikin do you maybe have a soap-ified VR catalog of a COLIBRE run to test all this?

EvgeniiChaikin commented:
Yes, I tested this branch before. 

But since there have been some new changes, I presume it is better to wait until cosma is back and then test this branch again on one of the most recent runs.

JBorrow commented:
Nice bug fix in the 1D velocity dispersion. Let me know when this is ready for merge.

EvgeniiChaikin commented:
I have done several tests. Here is one of them:

https://swift.dur.ac.uk/COLIBREPlots/chaikin/comparisons/YEAR2022/L50N376/z2.0_VR_vs_SOAP_[U102VR,U102SOAP]/index.html

where I compare VR and SOAP output made for the same simulation. 

There are about 4 plots where SOAP and VR still differ, but they probably all have to do with dust depletion. @james-trayford I have seen that you recently made some changes in the swift code, which may resolve the discrepancies we see here. Do you know if we need to also update anything in this PR that is related to dust depletion?

MatthieuSchaller commented:
I am happy to merge this too.

Just to confirm, if we really stop using VR altogether then we will not have to expand the big list in the translator object?

MatthieuSchaller commented:
> Do you know if we need to also update anything in this PR that is related to dust depletion?

Let's do this separately.

EvgeniiChaikin commented:
@MatthieuSchaller  I have tested SOAP again, now including the recent updates on the swift/colibre side from @james-trayford.

The gas metallicities plots now all look good to me. (You can see them yourselves here https://swift.dur.ac.uk/COLIBREPlots/chaikin/individual_runs/YEAR2022/z0.0_SOAP_TEST_MEATLS/#820371183054460454 where it is clear that metal depletion is accounted for)

However, while testing, I identified a new problem: the field with the linear average of Iron from SNIa over H, which was added relatively recently here https://github.com/SWIFTSIM/pipeline-configs/commit/06c0d1bf36349a18b89aaa9ba198017883f47d97, **is not registered** in SOAP

The pipeline errs with 


```
velociraptor-python/velociraptor/catalogue/translator.py", line 1402, in VR_to_SOAP
NotImplementedError: No SOAP analogue for property lin_element_ratios_times_masses.lin_FeSNIa_over_H_times_star_mass_50_kpc!
```

I overlooked this before because the SOAP branch of the pipeline configs wasn't up to date with the most recent master branch, so it did not include the update https://github.com/SWIFTSIM/pipeline-configs/commit/06c0d1bf36349a18b89aaa9ba198017883f47d97

To fix this, we need to add this one missing field to SOAP.

If I comment out this field, everything works perfectly!

EvgeniiChaikin commented:
@MatthieuSchaller  the commit I have just added fixes the error that I described above. 

I also created a pull request in the SOAP repo: https://github.com/SWIFTSIM/SOAP/pull/12, which creates the field that I have just put in the translator file.

With these two changes, the SOAP pipeline finally works!

MatthieuSchaller commented:
Just to confirm: You are happy with this to be merged?

EvgeniiChaikin commented:
Yes, I am happy if this is merged.

One thing to remember: https://github.com/SWIFTSIM/pipeline-configs/pull/233 and https://github.com/SWIFTSIM/SOAP/pull/12 need to be merged at the same time in order to not break anything.

JBorrow commented:
Can we merge this branch before merging those other two?

JBorrow commented:
I pushed a new release https://github.com/SWIFTSIM/velociraptor-python/releases/tag/v0.16.0
